processor/spanmetricsprocessor/processor.go

```diff
@@ -339,8 +339,13 @@ func (p *processorImp) getDimensionsByMetricKey(k metricKey) (*pcommon.Map, erro
 		}
 		return nil, fmt.Errorf("type assertion of metricKeyToDimensions attributes failed, the key is %q", k)
 	}
+	if item, ok := p.metricKeyToDimensions.EvictedItems[k]; ok {
```

Comment:
I thought `metricKeyToDimensions .Get(...)` already does this?

https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/0ea5dc49876c2298c06e235d22229ef0d4244c14/processor/spanmetricsprocessor/internal/cache/cache.go#L60

Reply:
yes, it do use the `evictedItems `, but still can't not explain why this error occur

Reply:
I debug for a while, I catch the point this time that:
https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/9810bb011e1edc835c8f8ae329884a183bf5586d/processor/spanmetricsprocessor/processor.go#L313-L324

It loops `p.callSum` containing all key-values since the collector is started to find a key in the cache, but the cache only has 1000 items by default. If the built different key is large than 1000, then the next loop of `p.callSum` will raise an error.

Reply:
Ah yes, I think you're right about the fact that `p.callSum` accumulates keys indefinitely since startup, unless if the user has configured delta temporality. 

For cumulative temporality, while I think we should still keep all keys since collector startup (so the counts are correct), I don't think we should loop over this entire set of metric keys. Instead we should only loop over those that relate to the current batch of spans received, by somehow marking a key as "dirty" if it relates to a span's set of metric keys.

What do you think?

I can put together some tests locally to test this theory.

Reply:
I think so, we just loop over those keys that relate to the current batch of spans received. Do I need close this PR and wait for your PR or I fix it in this PR?

Reply:
Please go ahead with the fix 👍🏼 

Note that https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/15710 involves a fairly major refactor of the spanmetrics processor, so I suggest to wait for https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/15710 to be merged first. 

As it's quite a large refactor, it might make more sense to close this PR and create a new one based off the refactored version.

Reply:
@Frapschen FYI https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/15710 has been merged.